### PR TITLE
[SDP-1276] Add retry for circle 429 requests

### PR DESCRIPTION
**Pull request by marwen-abid**
### What
Adding basic retry for Circle 429 retries

Circle requests in Sandbox environment usually fail after 6 requests. With this change we can finish processing all the payments in a disbursement. 

Screenshot 2024-07-23 at 7 25 58 PM

### Known limitations
* 🚨 Current implementation holds the database session until a retry is completed. In the case of Sandbox, that means the session is held for 5 minutes before the payment status is committed to the database. 

#### PR Structure

* [ ] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR title and description are clear enough for anyone to review it.
* [ ] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [ ] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [ ] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [ ] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [ ] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [ ] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [ ] This is not a breaking change.
* [ ] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: internal/circle/client.go [hunk 1]
```diff
@@ -177,22 +182,85 @@ func (client *Client) GetWalletByID(ctx context.Context, id string) (*Wallet, er
 	return parseWalletResponse(resp)
 }
 
+type RetryableError struct {
+	err        string
+	retryAfter time.Duration
+}
+
+func (err RetryableError) Error() string {
+	return err.err
```
Comment (marcelosalloum) [inline internal/circle/client.go:191]:
```suggestion
	return fmt.Errorf("retriable error: %w", err.err)
```

**Review by marcelosalloum — APPROVED** [review]
LGTM!

**Comment by marcelosalloum** [comment]
One additional thing we could do to prevent multiple Circle jobs from accumulating is to add a lock to CircleReconciliationService.Reconcile. If there's already a job being executed from that same IP, there's no point in starting another one.

https://github.com/stellar/stellar-disbursement-platform-backend/blob/6306eaddafc121f180ab9e0a271880db82da55ac/internal/services/circle_reconciliation_service.go#L33-L37

File: internal/circle/client.go (hunk 1 above)
Comment (marwen-abid) [reply internal/circle/client.go]: Addressed in 2bfdc05b0155ff373211ec3154658d97715633ae

**Pull request merged**